File: frontend/src/types/post.ts
export type PostStatus =
  | "draft"
  | "ready"
  | "scheduling"
  | "scheduled"
  | "failed"
  | "cancelled";

export type AttemptResult = "in_progress" | "success" | "failed";

export interface SchedulingAttempt {
  id: string;
  mode: "dry_run" | "facebook_schedule";
  result: AttemptResult;
  safe_message: string;
  error_code: string | null;
  external_request_made: boolean;
  created_at: string;
  completed_at: string | null;
}

export interface PostRecord {
  id: string;
  caption: string;
  image_url: string;
  image_mime_type: string;
  original_filename: string;
  status: PostStatus;
  scheduled_for_utc: string;
  scheduled_for_local: string;
  display_timezone: string;
  facebook_object_id: string | null;
  last_error_code: string | null;
  last_error_message: string | null;
  last_attempted_at: string | null;
  created_at: string;
  updated_at: string;
  attempts: SchedulingAttempt[];
}

export interface PostListResponse {
  items: PostRecord[];
  total: number;
}

export interface PostUpdateInput {
  caption?: string;
  scheduled_for_local?: string;
  timezone?: string;
}

export interface ScheduleResult {
  mode: "dry_run" | "facebook_schedule";
  simulated: boolean;
  success: boolean;
  post_id: string;
  attempt_id: string;
  post_status: PostStatus;
  external_request_made: boolean;
  facebook_object_id: string | null;
  message: string;
}

File: backend/app/schemas/post.py
"""Typed request and response contracts for persisted posts."""

from datetime import datetime

from pydantic import BaseModel

from app.models.post import AttemptResult, PostStatus, SchedulingMode


class SchedulingAttemptResponse(BaseModel):
    id: str
    mode: SchedulingMode
    result: AttemptResult
    safe_message: str
    error_code: str | None
    external_request_made: bool
    created_at: datetime
    completed_at: datetime | None


class PostResponse(BaseModel):
    id: str
    caption: str
    image_url: str
    image_mime_type: str
    original_filename: str
    status: PostStatus
    scheduled_for_utc: datetime
    scheduled_for_local: str
    display_timezone: str
    facebook_object_id: str | None
    last_error_code: str | None
    last_error_message: str | None
    last_attempted_at: datetime | None
    created_at: datetime
    updated_at: datetime
    attempts: list[SchedulingAttemptResponse]


class PostListResponse(BaseModel):
    items: list[PostResponse]
    total: int


class PostUpdateRequest(BaseModel):
    caption: str | None = None
    scheduled_for_local: str | None = None
    timezone: str | None = None


class ScheduleResponse(BaseModel):
    mode: SchedulingMode
    simulated: bool
    success: bool
    post_id: str
    attempt_id: str
    post_status: PostStatus
    external_request_made: bool
    facebook_object_id: str | None = None
    message: str


DryRunScheduleResponse = ScheduleResponse
